osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/MigratePlaylistScoresToSoloScoresCommand.cs

```diff
@@ -0,0 +1,190 @@
+                            {
+                                score.userId,
+                                score.roomId,
+                                score.beatmapId,
+                                score.playlistItemId,
+                                scoreId = insertId,
+                                createdAt = score.started_at,
+                                updatedAt = score.ended_at
+                            });
+
+                            Console.WriteLine();
+                        }
+                    }
+                }
+            }
+
+            Console.WriteLine("Finished.");
+            return 0;
+        }
+
+        private static HitResult getMaxJudgementFor(HitResult hitResult, HitResult max)
+        {
+            switch (hitResult)
+            {
+                case HitResult.Miss:
+                case HitResult.Meh:
+                case HitResult.Ok:
+                case HitResult.Good:
+                case HitResult.Great:
+                case HitResult.Perfect:
+                    return max;
+
+                case HitResult.SmallTickMiss:
+                case HitResult.SmallTickHit:
+                    return HitResult.SmallTickHit;
+
+                case HitResult.LargeTickMiss:
+                case HitResult.LargeTickHit:
+                    return HitResult.LargeTickHit;
+            }
+
+            return HitResult.IgnoreHit;
+        }
+
+        // ReSharper disable InconsistentNaming
+        public class MultiplayerScore
+        {
+            public long id { get; set; }
+            public uint userId { get; set; }
+            public long roomId { get; set; }
+            public long playlistItemId { get; set; }
+            public long beatmapId { get; set; }
+            public string rank { get; set; } = string.Empty;
+            public long? total_score { get; set; }
+            public double? accuracy { get; set; }
+            public double? pp { get; set; }
+            public int max_combo { get; set; }
+            public string mods { get; set; } = string.Empty;
+            public string statistics { get; set; } = string.Empty;
+            public DateTime started_at { get; set; }
+            public DateTime? ended_at { get; set; }
+            public bool passed { get; set; }
+            public DateTime? created_at { get; set; }
+            public DateTime? updated_at { get; set; }
+            public DateTime? deleted_at { get; set; }
+        }
+
+        public class MultiplayerPlaylistItem
+        {
+            public long id { get; set; }
+            public long room_id { get; set; }
+            public uint owner_id { get; set; }
+            public uint beatmap_id { get; set; }
+            public ushort ruleset_id { get; set; }
```

Comment:
Hmm in my DBs this is nullable / `ushort?`... Is production different? If not, may be worth running a

```sql
SELECT COUNT(1) FROM multiplayer_playlist_items WHERE ruleset_id IS NULL;
```

to check for any surprises there, because otherwise scores may accidentally get misattributed to osu!-the-ruleset... I don't have any rows with `NULL` in my dbs, but that means precisely nothing.

Reply:
I don't think it should be nullable.

```sql

root@db-master:osu> SELECT COUNT(*) FROM multiplayer_playlist_items WHERE ruleset_id IS NULL;
+----------+
| COUNT(*) |
+----------+
| 0        |
+----------+
```

Reply:
Okay great, just checking. Let's just say I'm being overly thorough in checking due to having an allergy for data anomalies from *prior bad experiences*...

Reply:
good catch, I believe this column can be just made `not null`. At least on osu-web side the field is required.

Reply:
(just gonna crosslink https://github.com/ppy/osu-web/pull/10523 for reference)

Reply:
Has been fixed.